src/net/server/handlers/mul.rs

```diff
@@ -0,0 +1,169 @@
+use crate::helpers::fabric::{ChannelId, MessageChunks, MessageEnvelope};
+use crate::helpers::Identity;
+use crate::net::server::{handlers::message_stream_layer::ReservedPermit, MpcServerError};
+use crate::net::RecordHeaders;
+use crate::protocol::{QueryId, RecordId, UniqueStepId};
+use async_trait::async_trait;
+use axum::extract::{self, FromRequest, Query, RequestParts};
+use axum::http::Request;
+use hyper::Body;
+
+/// Used in the axum handler to extract the `query_id` and `step` from the path of the request
+pub struct Path(QueryId, UniqueStepId);
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for Path {
+    type Rejection = MpcServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Path((query_id, step)) =
+            extract::Path::<(QueryId, UniqueStepId)>::from_request(req).await?;
+        Ok(Path(query_id, step))
+    }
+}
+
+#[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+pub struct IdentityQuery {
+    identity: Identity,
+}
+
+/// extracts the [`MessageChunks`] from the request and forwards it to the Message layer via the
+/// `permit`. If we try to extract the `permit` via the `Extension`'s `FromRequest` implementation,
+/// it will call `.clone()` on it, which will remove the `OwnedPermit`. Thus, we must access the
+/// `permit` via `Request::extensions_mut`, which returns [`Extensions`] without cloning.
+/// Unfortunately, this also means we must wrap the `permit` in an `Option` such that we can
+/// `take()` the `permit`, freeing up the `req` for `Request::into_body`
+#[allow(clippy::unused_async)] // handler is expected to be async
+#[allow(clippy::cast_possible_truncation)] // length of envelopes array known to be less u32
+pub async fn handler(
+    Path(_query_id, step): Path,
+    Query(IdentityQuery { identity }): Query<IdentityQuery>,
```

Comment:
can you please add a todo here to fix this once we figure out discovery and handshake stuff? Trusting client to tell us their identity is a big security hole and we shouldn't forget to fix it later. It is fine imo to have it right now, but would be great to have a list of things we need to fix

Reply:
done